오픈마켓 [STEP 1] 알라딘, 조이

junbangg opened the pull request:
---

안녕하세요 그린! **@GREENOVER🫒**

3조 알라딘과 조이입니다

`URLSessionTests` 를 진행하느라 API Request 메서드 구현은 끝까지 못했습니다ㅜㅜ

다음 스텝 진행할때 같이 구현해서 넣겠습니다!

말씀드린 일정을 꽉 채워 PR을 요청하게 되었네요ㅎㅎ

늦었지만 스텝1 잘 부탁드립니다😁

## 타입 소개

Protocols

- `DataTaskProvidable`: dataTask 메서드를 정의한 프로토콜 (MockURLSession 객체 주입을 위해 선언했습니다.)
- `APIServicable`: API 서비스를 지원을 위한 프로토콜
- `Parsable`: 파싱을 위한 프로토콜

Model

- `MarketAPIService`: 실제 네트워크 통신을 담당하는 클래스
- Page: 상품 리스트 조회를 위한 타입
- Product: 상품 타입
- PostProduct: (**POST / 상품 등록**)을 위한 타입
- PatchProduct: (**PATCH / 상품 수정**)을 위한 타입

## 테스트 소개

- `ParsableTests`: 파싱에 관한 테스트
- `APIServiceTests`: `MarketAPIService`의 메서드가 올바르게 동작하고 있는지 확인하는 테스트

# 고민했던 점

### 1. JSON Parsing을 하는 방법

- json 파싱을 담당하는 타입을 만들어줄지 아니면 파싱을 할 때마다 JSONDecoder를 생성해 줄지 고민했습니다.
- pop를 해보자는 생각이 들어서 `Parsable` 프로토콜을 정의한 후 기본메서드 구현에서 필요할때마다 JSONDecoder를 생성하는것으로 결정했습니다.

### 2. URLSession을 테스트 하는 방법

네트워킹을 실제로 안하면서 `APIService` 타입을 테스트하는 방법에 대해서 고민해봤습니다.

유닛 테스트는 실제 코드의 환경과는 독립 시키면서 빠르게 진행되어야하는데, 그러기 위해서는 실제 `APIService` 타입을 이용해서 단위테스트에서도 네트워킹을 하는게 아니라 `Mock` 객체를 만들어서 진행해야되는걸로 이해했습니다.

검색해보니까 `URLSession` 에 대한 **의존성 주입**을 통해 코드 구조를 바꿔서 테스트성을 높이는 흥미로운 방법이 있길래 이번 스텝에 그거에 대해 공부하고 적용해보는 시간을 가져봤습니다.

## 3. Protocol 사용

저희는 스텝 1부터, 구현해야되는 모든 기능을 `protocol` 로 먼저 정의 해보려고 노력했습니다.

`DataTaskProvidable` , `APIServicable` , `Parsable` 등, 프로그램에 필요한 기능이 생각나면 그걸 프로토콜로 먼저 정의해봤습니다.

# 조언을 얻고싶은 점

### 1.  status code의 정상범위를 명시하는 방법

status code 의 범위를 어디다가 정의 해놓고 사용할지에 대해 고민해봤습니다!

저희가 고민한 방법은 다음과 같습니다.

- 상수로 만들어놓고 클로져에서 사용하는 방법

```swift
let successRange = 200..<300
let dataTask = URLSession.shared.dataTask(with: request) {  data, response, error in
            guard error == nil,
                  let statusCode = (response as? HTTPURLResponse)?.statusCode,
                  successRange.contains(statusCode) else {
                      return
                  }
            
            guard let data = data else {
                completionHandler(.failure(APIError.unknownError))
                return
            }
            
            completionHandler(.success(data))
        }
```

- 범위를 그냥 명시하는 것

```swift
let dataTask = URLSession.shared.dataTask(with: request) {  data, response, error in
            guard error == nil,
                  let statusCode = (response as? HTTPURLResponse)?.statusCode,
                  (200..<300).contains(statusCode) else {
                      return
                  }
            
            guard let data = data else {
                completionHandler(.failure(APIError.unknownError))
                return
            }
            
            completionHandler(.success(data))
        }
```

1번처럼 범위를 상수로 빼줄지, 2번처럼 그냥 범위를 적어줄 지 고민했습니다. 저희는 코드중간에 숫자가 날것으로 존재하는것을 참을 수 없어서 1번 방법을 선택했어요. 

위 경우에서는 직접 범위를 명시하는게 좀 더 가독성이 좋을지 그린의 의견이 궁금합니다.🙋🏻‍♀️

### 2 `unowned` vs `weak`사용법.

```swift
let dataTask = session.dataTask(with: request) { [weak self] data, response, error in
    guard error == nil else {
        completionHandler(.failure(APIError.invalidHTTPMethod))
        return
    }
    guard let statusCode = (response as? HTTPURLResponse)?.statusCode,
          successRange.contains(statusCode) else {
        completionHandler(.failure(APIError.invalidRequest))
        return
    }
    guard let data = data else {
        completionHandler(.failure(APIError.noData))
        return
    }
    guard let parsedData = self?.parse(with: data, type: T.self) else {
        return
    }
    
    completionHandler(.success(parsedData))
}
```

위 코드에서 `self` (`MarketAPIService`) 를 클로져에서 접근하기 위해서 `weak self` 로 했습니다.

그런데 생각해보니까 `self`(`MarketAPIService`)가 클로저보다 생명주기가 길것 같은데 이럴때는 `unowned self` 를 사용하는게 맞는걸까요??

저희가 올바르게 고민한것인지 궁금합니다.

## 3. API URL Namespace 에 대한 조언

`API` 관련한 문자열들을 `Namespace` 에 이쁘게 모아두려고 노력해봤습니다.

`var url: URL? {}` 이라는 연산 프로퍼티를 이용해서 아예 `URL` 타입을 리턴 하는 방법으로 구현을 해봤는데

저희가 사용한 방법에 대해서 어떻게 생각하시는지, 또는 조언 주실만한 추가적인 방법이 있을지 궁금합니다🙏

na-young-kwon committed:
feat: Product 타입 구현 #1

junbangg committed:
feat: Page 타입 구현 #1

na-young-kwon committed:
feat: URLSessionProtocol 추가 #1

junbangg committed:
feat: APIServiceable 프로토콜 구현 #1

- PostProduct, PatchProduct 타입 구현

na-young-kwon committed:
feat: APIServicable extension의  parse 메서드 구현 #1

junbangg committed:
feat: 단위 테스트 환경 구축 및 Parsable 프로토콜 추가 #1 #2

na-young-kwon committed:
test: Page를 파싱하는 메서드에 대한 테스트 케이스 추가 #2

junbangg committed:
test: Product 모델 검증하는 테스트 작성 #2

- Product에 Equatable 프로토콜 적용
- parsePage 메서드 추가

na-young-kwon committed:
test: parse 메서드를 검증하는 테스트 작성 #2

- parse 메서드 제네릭 타입으로 변경
- JSONError 타입 추가
- products의 마지막 요소를 검증하는 테스트 케이스 추가

junbangg committed:
feat: 상품 상세 조회를 처리하는 메서드 구현 #1

- APIError 타입 추가

na-young-kwon committed:
feat: 상품 리스트 조회를 처리하는 메서드 구현 #1

junbangg committed:
refactor: MarketAPIService 리팩토링 #1

- URL Namespace 생성
- URLDataTask 생성하는 코드 메서드로 분리

na-young-kwon committed:
refactor: parse 메서드가 에러를 던지지 않도록 수정 #1

- parse 메서드 호출부에서 catch 부분 삭제
- MarketAPIService 타입 class -> final class 로 변경

junbangg committed:
test: URLSession 단위 테스트를위한 테스트 더블 구현 #2

- MockURLSession 구현
- MockURLSessionDataTask 구현

na-young-kwon committed:
refactor: createDataTask메서드 반환타입 및 메서드명 변경#1

- parse메서드가 optional을 반환하도록 변경
- performDataTask 메서드에서 dataTask를 수행하도록 로직 수정
- 상품리스트조회 기능의 잘못된 url내용 수정

junbangg committed:
refactor: 프로토콜 이름 변경 및 테스트 코드 리팩토링 #1 #2

- URLSessionProtocol 에서 DataTaskProvidable 으로 변경
- parse 메서드 로직 변경에 변화에 맞게 테스트 코드 변경
- 테스트 환경 실제 코드와 분리

na-young-kwon committed:
test: MarketAPIService의 메서드를 검증하는 테스트케이스 추가 #1 #2

- 가짜 데이터 주입을 위한 MockRequest 열거형 추가
- APIError LocalizedError -> Error로 변경
- 코딩 컨벤션준수를 위해 makeRequestFail -> isSuccessfulRequest 로 변경

junbangg committed:
test: getProduct에 대한 단위테스트 작성 #2

- JSONError 타입 LocalizedError 에서 Error으로 변경

na-young-kwon committed:
refactor: successRange 프로퍼티 메서드 안으로 위치 변경 #1

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -319,6 +517,48 @@
+				CODE_SIGN_STYLE = Automatic;
+				DEVELOPMENT_TEAM = 7QHGBDS383;
+				INFOPLIST_FILE = OpenMarketTests/Info.plist;
+				IPHONEOS_DEPLOYMENT_TARGET = 14.5;
```
OpenMarket의 미니멈 타겟은 13.2 버전인데 테스트로는 14.5 버전을 미니멈 타겟으로 가져가신 이유가 궁금합니다🙋🏻

junbangg replied:
음 그건 저희가 생각을 못했던것 같습니다,,,:) 고쳐놓겠습니다!

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -7,15 +7,48 @@
 		C70FB10425BEF61D00C9924E /* Assets.xcassets in Resources */ = {isa = PBXBuildFile; fileRef = C70FB10325BEF61D00C9924E /* Assets.xcassets */; };
 		C70FB10725BEF61D00C9924E /* LaunchScreen.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = C70FB10525BEF61D00C9924E /* LaunchScreen.storyboard */; };
+		ED0157E52782E5C400BEE737 /* Page.swift in Sources */ = {isa = PBXBuildFile; fileRef = ED0157E42782E5C400BEE737 /* Page.swift */; };
+		ED6EEFF927870D750042B6C2 /* Int+Extensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = ED6EEFF827870D750042B6C2 /* Int+Extensions.swift */; };
```
기본 자료형의 Int의 extension 파일인것 같은데 파일 위치 구조를 보면 Model > Extenstion인데
Model 하위로 가져가신 이유가 궁금합니다🙋🏻
파일 구조에 정답은 없지만 이 경우에는 기본 자료형일거라 Model이 아니더라도 추후 두루 쓰일 수 있을것 같은 부분이라고 생각이 들었어요!
그렇다면 Model과 동일한 선상의 레벨 단위에서 그룹핑을 해주면 어떨까 합니다🤔

na-young-kwon replied:
저희가 실수로 파일 위치를 잘못잡은것 같아요! 
위에서 말씀해주신대로 Model이 아니더라도 추후 두루 쓰일 수 있을것같아서 Model과 동일한 선상으로 이동시켜주었습니다.

GREENOVER commented on OpenMarket/OpenMarket/Model/Errors/JSONError.swift:
```diff
@@ -0,0 +1,21 @@
+extension JSONError {
+    var description: String {
+        switch self {
+        case .parsingError:
```
조금 더 명확한 `parsingFail` 같은 네이밍은 어떨까요?
JSONError.parsingError로 본다면 Error라는 정보가 중첩되는것 같습니다🤔

junbangg replied:
`parsingFail` 이 훨씬 좋은것 같습니다!
근데 아래에서 그린이 답글 주신 내용 보고, 이 `JSONError` 는 필요가 없는것 같아서 삭제하기로 했습니다!

GREENOVER commented on OpenMarket/OpenMarket/Model/Errors/JSONError.swift:
```diff
@@ -0,0 +1,21 @@
+        case .parsingError:
+            return "JSON 파싱을 실패했습니다."
+        }
+    }
```
case에 String으로 바로 넣어줄 수도 있을텐데 computed property로 구현을 택해주신 이유가 궁금합니다~!🙋🏻
(잘못되었다는거 절대 절대 아닙니다!)

na-young-kwon replied:
Enum의 RawValue를 특정 값을 저장하려는 목적으로 사용하는것이 올바르지 않다고 생각했기 때문에 연산프로퍼티로 구현해주었습니다. RawValue는 case 생성을 위한 값이기 때문에  Identifier 같은 의미있는 값을 사용해야 한다고 생각합니다. 🙋🏻‍♀️

GREENOVER commented on OpenMarket/OpenMarket/Model/Errors/APIError.swift:
```diff
@@ -0,0 +1,15 @@
+    case invalidHTTPMethod
+    case invalidRequest
+    case noData
+}
```
case 네이밍이 명확해서 좋네요!👍
요 에러들은 자세한 description이 필요없을까요? 
있으면 좋을듯 하다고 판단되서요..!🤔

na-young-kwon replied:
description을 쓴다고 했는데 깜빡했네요,,ㅎㅎ
수정해주었습니다!!

GREENOVER commented on OpenMarket/OpenMarket/Model/Protocols/DataTaskProvidable.swift:
```diff
@@ -0,0 +1,15 @@
+import Foundation
+
+protocol DataTaskProvidable {
+    func dataTask(with request: URLRequest,
+                      completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask
```
파라미터 시그니쳐가 길어져서 가독이 조금 떨어지는데 이런 코드 스타일은 어떠신가요?🙋🏻
```swift
func dataTask(
    with request: URLRequest,
    completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void
) -> URLSessionDataTask
```
실제로 가로로 긴것보다 세로로 긴게 더 개발자들이 선호하는 스타일로 알고 있고 Swift에서도 위와 같이 스타일을 가져가는걸로 알고 있어요..!
위 코드 스타일의 장점은 파라미터 시작과 종료를 더 명확하게 구조 파악하는데 도움이 됩니다!

junbangg replied:
오 딱 봐도 더 깔끔하네요!!
그렇게 수정해보겠습니다~~

GREENOVER commented on OpenMarket/OpenMarket/Model/Protocols/Parsable.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+protocol Parsable {
```
어떤 Parsable인지 조금 더 명확한 네이밍이면 좋을것 같습니다🙋🏻

junbangg replied:
수정했습니다!

na-young-kwon replied:
`JSONParsable`로 수정해주었습니다 :)

GREENOVER commented on OpenMarket/OpenMarket/Model/Protocols/Parsable.swift:
```diff
@@ -0,0 +1,27 @@
+            print(JSONError.parsingError.description)
+        } catch {
+            print(error.localizedDescription)
+        }
```
두번의 catch는 각 어떤 에러를 캐치해주는건가요~?🙋🏻

na-young-kwon replied:
첫번째 catch문은 JSONError.parsingError 를 캐치합니다.
(원래 해당 메서드가 에러를 throw하고 있었는데, nonThrowing 메서드로 변경하면서 던지지도 않는 에러를 catch 하고있네,,,,요. 이 에러는 더 이상 사용되는 곳이 없어서 삭제해주었습니다~!)

두번째 catch문은 `JSONDecoder().decode(type, from: data)`에서 docode를 할 때 발생하는 오류를 캐치합니다.

GREENOVER commented on OpenMarket/OpenMarket/Model/Protocols/APIServicable.swift:
```diff
@@ -0,0 +1,18 @@
+    func post(productID: Int, secret: String)
+    func delete(productID: Int, productSecret: String)
+    func get(productID: Int, completionHandler: @escaping (Result<Product, APIError>) -> Void)
+    func get(pageNumber: Int, itemsPerPage: Int, completionHandler: @escaping (Result<Page, APIError>) -> Void) 
```
같은 메서드를 오버라이딩하고 파라미터로 구분지어주면 나중에 구현하실때 혼동이 없을까요?🤔
코드가 많아지고 복잡해지면 실수할 여지가 있어보입니다🥲

junbangg replied:
각 메서드 네이밍 수정했습니다!

GREENOVER commented on OpenMarket/OpenMarket/Model/Page.swift:
```diff
@@ -0,0 +1,32 @@
+        case pageNumber = "page_no"
+        case itemsPerPage = "items_per_page"
+        case totalCount = "total_count"
+        case offset
+        case limit
```
변환이 없는 프로퍼티들은 한줄로 표현해도 좋겠네요🤔
```swift
case offset, limit
```
이러한 프로퍼티들이 많아지면 코드가 증가하기에 한줄로 표현하는걸 저는 선호합니다!
정말 그런데 스타일 차이긴해서 이런것도 있다는걸 보시면 될것 같아요ㅎㅎ

na-young-kwon replied:
이런방법도 있었군요.! 프로퍼티들이 많아지면 한줄로 적어주는것이 가독성에 좋을 것 같아요! 수정해주었습니다 :)

GREENOVER commented on OpenMarket/OpenMarket/Model/PostProduct.swift:
```diff
@@ -0,0 +1,28 @@
+        case descriptions
+        case price
+        case currency
+        case discountedPrice = "discounted_price"
```
아마 아실수도 있는데 만약 스네이크 케이스에서 단순히 카멜 케이스로만 바꿔주는 코딩키 전략을 세우시게 된다면
keyDecodingStrategy와 keyEncodingStrategy와 같은 유용한 프로퍼티도 존재합니다~!
https://developer.apple.com/documentation/foundation/jsondecoder/2949119-keydecodingstrategy
https://developer.apple.com/documentation/foundation/jsonencoder/2949141-keyencodingstrategy/
여러분들의 프로젝트에서 사용되지 않아도되요!
어떤게 더 편리하고 유용한지 판단해보면 될것 같아요ㅎㅎ
제 판단에는 지금 모델 타입에서는 단순 네이밍 스타일 컨벤션 외에 아예 네이밍 자체가 변하는것도 많아서 소개해드린 방식은 적합하진 않아 보이긴 합니다!

na-young-kwon replied:
넵! 링크 감사합니다. 현재 프로젝트에서는 네이밍 자체가 변하는 경우가 있어서 이대로 진행할게요!

GREENOVER commented on OpenMarket/OpenMarket/Model/Product.swift:
```diff
@@ -0,0 +1,36 @@
+
+import Foundation
+
+struct Product: Decodable {
```
오호 response에 해당하는 친구들은 Decodable로 request에 해당하는 친구들은 Encodable로 가셨군요!
명확하네요👍

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -0,0 +1,125 @@
+//MARK: - Namespace
+
+extension MarketAPIService {
+    private enum MarketAPI {
+        static let baseURL = "https://market-training.yagom-academy.kr"
+        static let path = "/api/products/"
+        
+        case postProduct
+        case patchProduct(id: Int)
+        case postSecret(id: Int)
+        case delete(id: Int, secret: String)
+        case getProduct(id: Int)
+        case getPage(pageNumber: Int, itemsPerPage: Int)
+        
+        var url: URL? {
+            switch self {
+            case .postProduct:
+                return URL(string: MarketAPI.baseURL + MarketAPI.path)
+            case .patchProduct(let id):
+                return URL(string: MarketAPI.baseURL + MarketAPI.path + id.stringFormat)
+            case .postSecret(let id):
+                return URL(string: MarketAPI.baseURL + MarketAPI.path + id.stringFormat + "/secret")
+            case .delete(let id, let secret):
+                return URL(string: MarketAPI.baseURL + MarketAPI.path + id.stringFormat + "/" + secret)
+            case .getProduct(let id):
+                return URL(string: MarketAPI.baseURL + MarketAPI.path + id.stringFormat)
+            case .getPage(let id, let itemsPerPage):
+                guard var urlComponents = URLComponents(string: MarketAPI.baseURL + MarketAPI.path) else {
+                    return nil
+                }
+                let pageNumberQuery = URLQueryItem(name: "page_no", value: id.stringFormat)
+                let itemsPerPageQuery = URLQueryItem(name: "items_per_page", value: itemsPerPage.stringFormat)
+                
+                urlComponents.queryItems = [pageNumberQuery, itemsPerPageQuery]
+                
+                return urlComponents.url
+            }
+        }
+    }
+}
```
MarketAPI를 MarketAPIService에 extension하지 않고 별도 파일과 타입으로 분리할 수도 있을텐데
extension을 활용하신 이유가 궁금합니다~!🙋🏻
API 자체를 marketAPI로 국한하지 않고 추후 다른 API들이 들어올 여지를 감안해 공통적인 API enum 타입을 가지고 활용해보면 어떨까해서요..!
또 궁금한점이 있습니다🤔
코드 전역적으로 extension을 적극 활용해주셨는데 여러분들은 어떤 기준으로 extension을 사용하시는지 궁금해요🙋🏻

junbangg replied:
저희는 `MarketAPI` 타입을 `MarketAPIService` 내부에서만 사용할것 같아서 중첩타입으로 만들었는데, 말씀해주신것처럼 공통적으로 사용할 수 있는 API 타입을 만드는 방법도 있겠군요! 그거에 대해서는 생각을 안해봤습니다..
공통적으로 사용할 수 있는 API 타입을 만드는 방법에 대해서 고민을 해봤는데, 방법이 잘 떠오르지않습니다.. 밖으로 빼도 결국 `API` 별로 분리가 될것 같은 생각이 드는데, 어떤 식으로 관리하는건지 궁금합니다!

그리고 저희가 `extension` 을 활용한 기준은, 타입의 **구성요소**들(저장속성, init, 메서드, 중첩타입, 프로토콜 채택)을 정리하기 위함이었습니다. `MarketAPIService` 또는 `ViewController` 처럼 한 타입에 많은 코드가 있을 경우에 가독성이 안좋아질 수 있어서 `extension` 으로 관련된 **구성요소**들끼리 묶어서 구분지으면 조금이라도 더 깔끔하고 가독성이 좋아진다는 생각을 했습니다.
어떻게보면 `extension`을 타입 내부에서 적절한 색인의 역할로 생각하고 사용했습니다:)

GREENOVER replied:
맞습니다!
밖으로 빼도 결국 API 별로 분리가 될것 같다는게 API별로 여러 Enum 타입이 생길 수 있음을 말씀하시는걸까요?
그렇다면 그건 Enum 타입으로 만들고 호출 시 인자를 넘겨줘 해결할 수 있을것 같습니다.
제가 여기서 말씀드렸던 포인트는 답변 달아주신 첫문장과 동일합니다!
저 Enum을 정말 내부에서만 쓰는거면 지금처럼 가셔도 됩니다.
다만 추후에 저런 형식을 다양하게 쓰게 될 여지가 있다면 외부로 분리해보는게 어떨까 하는 의견을 드렸고
결국 두 방법 모두 정답이며 두분의 의견이 궁금했습니다😀

junbangg replied:
호출시 인자로 넘겨서 해결할 수 있겠군요?!?그런 방법에 대해서는 생각을 못해봤는데, 앞으로 스텝 진행하면서 계속 생각을 해보겠습니다:)

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -0,0 +1,125 @@
+final class MarketAPIService {
+    private let session: DataTaskProvidable
+    
+    init(session: DataTaskProvidable = URLSession.shared) {
+        self.session = session
+    }
```
이니셜라이저를 해준 이유가 있을까요?🙋🏻
어느 기준을 가지고 이니셜라이저를 구현해주시나요~?
저 session이 URLSession.shared의 값이 아닐때가 있는건가요🤔

junbangg replied:
의존성 주입을 위해서 `session` 저장 프로퍼티에 기본 값을 안주려고 했습니다.
추가적으로, 이렇게 `URLSession` 프로퍼티와 `MarketAPIService` 타입간의 결합도를 낮춤으로서, 테스성을 높일 수 있었다고 생각합니다. 
그래서 단위테스트를 할때도, `StubURLSession` 을 `MarketAPIServce` 에 주입해서 테스트할 수 있게 됐습니다.

`init` 내부에 `URLSession.shared` 기본값을 준것은, `MarketAPIService` 를 생성할때, 따로 `URLSession` 을 주입 안해도 되는 옵션을 주기 위함입니다.

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -0,0 +1,125 @@
+    }
+}
+
+//MARK: - Namespace
```
Namespace로 다뤄주신거 아주아주 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -0,0 +1,125 @@
+
+extension MarketAPIService {
+    private func performDataTask<T: Decodable>(request: URLRequest,
+                                              completionHandler: @escaping (Result<T, APIError>) -> Void) {
```
요 부분도 조금 더 가독성을 높여볼 수 있을것 같아요🙋🏻

na-young-kwon replied:
위에서 말씀해주신것 처럼 파라미터 길어지는 메서드는 세로로 작성해주었습니다. 
파라미터의 시작과 종료가 좀 더 명확해진것같아요👍🏻

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -0,0 +1,125 @@
+extension MarketAPIService {
+    private func performDataTask<T: Decodable>(request: URLRequest,
+                                              completionHandler: @escaping (Result<T, APIError>) -> Void) {
+        let successRange = 200..<300
```
상수 처리에 한표요~!👍

GREENOVER commented on OpenMarket/OpenMarket.xcodeproj/project.pbxproj:
```diff
@@ -7,15 +7,48 @@
+		ED0157E52782E5C400BEE737 /* Page.swift in Sources */ = {isa = PBXBuildFile; fileRef = ED0157E42782E5C400BEE737 /* Page.swift */; };
+		ED6EEFF927870D750042B6C2 /* Int+Extensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = ED6EEFF827870D750042B6C2 /* Int+Extensions.swift */; };
+		ED6EEFFB278728E90042B6C2 /* MockURLSession.swift in Sources */ = {isa = PBXBuildFile; fileRef = ED6EEFFA278728E90042B6C2 /* MockURLSession.swift */; };
+		ED6EF00D2788320F0042B6C2 /* Page+Extensions.swift in Sources */ = {isa = PBXBuildFile; fileRef = ED6EF00C2788320F0042B6C2 /* Page+Extensions.swift */; };
```
익스텐션 해줄때 파일 네이밍이 아주 좋네요!
저도 이렇게 해요👍
많은 외부라이브러리들이 익스텐션 시 이런 네이밍 규칙을 따르는걸로 알고 있습니다!

GREENOVER commented on OpenMarket/OpenMarketTests/Test Extensions/Page+Extensions.swift:
```diff
@@ -0,0 +1,23 @@
+
+extension Page: Equatable {
+    public static func == (lhs: Page, rhs: Page) -> Bool {
+        return lhs.pageNumber == rhs.pageNumber &&
+            lhs.itemsPerPage == rhs.itemsPerPage &&
+            lhs.totalCount == rhs.totalCount &&
+            lhs.offset == rhs.offset &&
+            lhs.limit == rhs.limit &&
+            lhs.products == rhs.products &&
+            lhs.lastPage == rhs.lastPage &&
+            lhs.hasNext == rhs.hasNext &&
+            lhs.hasPrevious == rhs.hasPrevious
```
전부다 비교해주는 이유가 궁금합니다~!🙋🏻

na-young-kwon replied:
처음 테스트코드를 작성할 때 어떤식으로 비교를 해줄지 고민했습니다. 
`XCTAssertEqual(20, testProduct.id)` 이런식으로 비교하는 방법도 생각해보았는데
이렇게 비교하는것 보다 인스턴스끼리 비교를 하는것이 좀 더 적합한 테스트라고 생각했습니다. (우연히 testProduct의 id가 20일 수 있으니까요!)

이 때 인스턴스끼리 비교를 해주려면 비교해주려는 타입이 Equatable을 채택해야 하고, `타입의 모든 프로퍼티가 동일한지 테스트를 하자`는 목적에 맞게 전부 다 비교를 해주고 있습니다.

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLSession.swift:
```diff
@@ -0,0 +1,48 @@
+import UIKit.NSDataAsset
+@testable import OpenMarket
+
+final class MockURLSessionDataTask: URLSessionDataTask {
```
요 타입은 init() 사용이 iOS 13부터 deprecated 되었는데 추후 이니셜라이저의 사용이 필요한 경우는 없을까요?🤔

junbangg replied:
deprecated warning을 어떻게 해결하는게 좋을까요???
저희도 mock 객체를 생성하기위한 이니셜라이저가 필요한데 말이죠🥲
자료들을 참고해봤는데 저희가 필요한 `URLSessionDataTask.resume()` 를 `override` 하는 방법을 모르겠습니다.

GREENOVER replied:
이 경우에는 URLSession을 상속받는 커스텀한 Stub(Mock)URLSession을 구현함으로 해결할 수 있습니다.
찾아보니 WWDC2018에서 소개를 했었더라구요..!
https://developer.apple.com/videos/play/wwdc2018/417/
![스크린샷 2022-01-11 오전 8 54 51](https://user-images.githubusercontent.com/72292617/148857617-22989eae-b983-4df3-9edf-047dbb8a0a78.png)
위 사진은 How To URL Protocol 섹션을 소개할때 나옵니다.
너무 명확하게 어떻게 사용하는지 잘 나오네요..!
저 4개의 메서드를 오버라이드하여 사용하면 됩니다.
물론 보시기에 개념이 쉽진 않을 수 있습니다!
그래도 저 방법을 가장 많이 대체하여 테스트를 사용하고 있습니다~!

junbangg replied:
오 감사합니다!! wwdc가 있었군요..참고하겠습니다! 🙏

na-young-kwon replied:
좋은 자료 감사합니다
영상 참고해서 테스트를 바꿔주려고 했는데, 생각보다 시간이 오래걸릴것 같다는 생각이 들었어요! (테스트 형식등을 다 바꿔줘야 하더라구요ㅜ)
step2에서 컬렉션뷰를 경험해보는것도 중요하다는 생각이 들어서 step2 먼저 진행하고 추후에 개선해보려고 합니다 :)

GREENOVER replied:
@na-young-kwon 
네 좋습니다!
스텝2에서 요구되는 사항부터 처리하고 추후 여유가될때 이 부분은 해봐도 무방할것 같습니다!

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLSession.swift:
```diff
@@ -0,0 +1,48 @@
+
+
+final class MockURLSession: DataTaskProvidable {
+    var isSuccessfulRequest: Bool
```
이 프로퍼티는 변경될 여지가 있나요?🤔

na-young-kwon replied:
변경되지 않는 값이라서 상수로 바꿔주었습니다 :O

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLSession.swift:
```diff
@@ -0,0 +1,48 @@
+final class MockURLSession: DataTaskProvidable {
+    var isSuccessfulRequest: Bool
+    var sessionDataTask: MockURLSessionDataTask?
+    var request: MockRequest
```
요 프로퍼티도 변경될 여지가 있나요?🤔

na-young-kwon replied:
request도 상수로 변경해주었습니다

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLSession.swift:
```diff
@@ -0,0 +1,48 @@
+        sessionDataTask.resumeDidCall = {
+            if self.isSuccessfulRequest {
+                completionHandler(self.request.data, successResponse, nil)
+                
```
여기 줄바꿈의 의미는 어떤걸까요?🙋🏻

na-young-kwon replied:
실수였어요ㅎㅎㅎ 
삭제했습니다~!

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockRequest.swift:
```diff
@@ -0,0 +1,85 @@
+    """.utf8
+    
+    static let getProductData = """
+            {
+                "id": 87,
+                "vendor_id": 3,
+                "name": "aladdin",
+                "thumbnail": "https://s3.ap-northeast-2.amazonaws.com/media.yagom-academy.kr/training-resources/3/thumb/b07b4b206eec11ecabfa73cd0451c065.jpeg",
+                "currency": "KRW",
+                "price": 0.0,
+                "bargain_price": 0.0,
+                "discounted_price": 0.0,
+                "stock": 0,
+                "created_at": "2022-01-06T00:00:00.00",
+                "issued_at": "2022-01-06T00:00:00.00",
+                "images": [
+                    {
+                        "id": 60,
+                        "url": "https://s3.ap-northeast-2.amazonaws.com/media.yagom-academy.kr/training-resources/3/origin/b07b4b206eec11ecabfa73cd0451c065.jpeg",
+                        "thumbnail_url": "https://s3.ap-northeast-2.amazonaws.com/media.yagom-academy.kr/training-resources/3/thumb/b07b4b206eec11ecabfa73cd0451c065.jpeg",
+                        "succeed": true,
+                        "issued_at": "2022-01-06T00:00:00.00"
+                    }
+                ],
+                "vendors": {
+                    "name": "Vendor2",
+                    "id": 3,
+                    "created_at": "2021-12-27T00:00:00.00",
+                    "issued_at": "2021-12-27T00:00:00.00"
+                }
+            }
+            """.utf8
```
이 부분의 indent가 위와 상이한것 같아요🧐

na-young-kwon replied:
마지막 `"""` 의 위치에 따라서 인덴트가 바뀌는군요!
알라딘이 한땀한땀 수정해주었습니다ㅋㅋㅋㅋㅋㅋ

GREENOVER commented on OpenMarket/OpenMarketTests/Tests/ParsableTests.swift:
```diff
@@ -0,0 +1,105 @@
+@testable import OpenMarket
+
+final class ParsableTests: XCTestCase {
+    var sut: Parsable!
```
force Unwrapping을 하지 않는 방향은 어떨까요?🙋🏻
테스트 코드여서 괜찮긴 하지만 사실 대부분의 앱 크래쉬의 주요 원인은 force Unwrapping일때가 있습니다😭

junbangg replied:
그렇군요! 처음 배웠을때 저렇게 하는 경우만 봐왔는데 그럴 수도 있겠네요!
옵셔널로 바꿔봤습니다!

GREENOVER commented on OpenMarket/OpenMarketTests/Tests/ParsableTests.swift:
```diff
@@ -0,0 +1,105 @@
+final class ParsableTests: XCTestCase {
+    var sut: Parsable!
+    
+    override func setUpWithError() throws {
+        sut = MarketAPIService()
+    }
+
+    override func tearDownWithError() throws {
+        sut = nil
+    }
```
LGTM👍

GREENOVER commented on OpenMarket/OpenMarketTests/Test Doubles/MockURLSession.swift:
```diff
@@ -0,0 +1,48 @@
+    func dataTask(with request: URLRequest, completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void) -> URLSessionDataTask {
+        let successResponse = HTTPURLResponse(url: request.url!, statusCode: 200, httpVersion: nil, headerFields: nil)
+        let failureResponse = HTTPURLResponse(url: request.url!, statusCode: 410, httpVersion: nil, headerFields: nil)
+        let sessionDataTask = MockURLSessionDataTask()
```
위 deprecated 관련 이어지는 부분이긴 합니다~!
이 부분 컴파일 시 warning message가 뜨는데 빌드에 지장은 없지만 우선 경고 메시지를 무시하면 나중에 큰 화가 됩니다ㅠ
어떻게 하면 좋을까요?🤔

na-young-kwon replied:
위에서 코멘트남겼는데 개선하는 방법을 잘 모르겠습니다,,

GREENOVER replied:
위에 코멘트 남겨봤습니다🙌

GREENOVER commented on OpenMarket/OpenMarketTests/Tests/ParsableTests.swift:
```diff
@@ -0,0 +1,105 @@
+import XCTest
+@testable import OpenMarket
+
+final class ParsableTests: XCTestCase {
```
어떤 ParsableTests인지 조금 더 명확하면 베스트일것 같아요🙋🏻

na-young-kwon replied:
JSONParsableTests 로 변경해주었습니다!

GREENOVER reviewed (changes requested):
@junbangg @na-young-kwon 
안녕하세요 알라딘 조이 !!🙇🏻
이번 2주간 여러분과 오픈마켓 프로젝트를 같이 진행할 그린입니다ㅎㅎ
올리브를 붙여주셨네요🫒!!ㅋㅋ
늦었지만 새해복 많이 받으세요🙇🏻
제 리뷰를 검사보다는 동료 개발자에게 피드백 받는 그런 시간임으로 제 리뷰가 타당하지 않거나 다른 의견이 있다고 느껴지는 부분은
수정하기보다는 꼭 의견을 남겨주세요👍
그럼 2주간의 협업 잘 부탁드립니다🥳
***
### 칭찬 드리고 싶은 부분
- 깃헙의 이슈 사용을 잘 해보셨군요 💯
- 명확하고 잘 나눠진 커밋 메시지가 일품이네요 💯
- 전체적으로 파일을 목적에 맞게 잘 나누셨어요 💯
- extension과 protocol의 적극적인 활용 💯
***
### 고민하셨던점에 대한 저의 의견
1. JSON Parsing을 하는 방법
- 잘 구현되었고 실제로 코드 동작에도 이상없을겁니다!
다만 궁금한게 JSON 파싱을 담당하는 타입을 따로 만드시려고 생각은 하셨는데 그렇게 진행되면 어떤 이점이 있을까요?
그리고 `pop을 해보자는 생각이 들어서`라는 문구를 제가 잘 이해하지 못했습니다🥲
결론적으로 타입을 따로 가져가시지 않은 선택의 이유가 궁금합니다🙋🏻
2. URLSession을 테스트 하는 방법
- 테스트 코드 잘 구현되었네요!
그거와 별개로 원론적인 Test Doubles에서의 개념에 대해 얘기를 해볼까 합니다.
여러분은 주로 mock의 테스트를 진행해주신것 같아요.
mock 객체를 생성하여 이를 사용하는건 맞지만 요구사항에 나오는 네트워크를 실제로 하지 않으면서 이와 동일한 결과값을 도출해내고
이러한 테스트는 어떤 종류의 테스트 일까요?
이러한 테스트는 stub 테스트로 많이 불리곤 합니다.
mock, stub, fake 등 테스트의 종류가 많은데 사실 네이밍 붙이기 나름이고 그 안에서 로직을 짜기 나름이지만
현업에서 개발할 시 저렇게 명시적으로 테스트 종류를 구분 지은 이유가 있습니다.
각 특성이 있고 이를 지켜서 원하는 테스트를 나눌 수 있어요.
아마 오동나무도 야곰 아카데미 테크 코스에 비교적 최근에 TestDouble에 대해 강좌를 올려 놓았던것 같아요.
저도 슬~쩍 봤는데 되게 잘 되어 있습니다!
그와 별개로 저는 Test Double에 대해 개념을 잡을때 이 레퍼런스를 참고 했어요.
https://blog.pragmatists.com/test-doubles-fakes-mocks-and-stubs-1a7491dfa3da
개념을 잡기에는 충분하실거에요!
이를 토대로 지금 구현하시려는게 어떤 종류의 테스트이며 왜 필요한지 이해하시는것을 추천드립니다🙋🏻
3. Protocol 사용
- 여러분들 코드에서 프로토콜 사용이 적절히 잘 이뤄졌다고 생각해요👍
이에 별다른 코멘트는 생각나지 않고 네이밍과 구조에 대해서만 리뷰 코멘트로 달아봤어요ㅎㅎ
여기에 잠깐 얘기해보자면 protocol 그룹이 Model 하위 그룹에 속해있어서 조금 어색하다고 느껴지긴 했습니다!
이건 정말 스타일 차이긴해서..
근데 여러분들의 의도는 해당 프로토콜들은 어차피 모델에서만 국한되기에 이렇게 파일 구조를 잡으셨겠구나 라고 짐작했는데 맞을까요?
아 보통 저는 Model에만 국한되지 않을 여지가 있다보니 Model / Extensions / Errors / Protocols를 같은 레벨 선상으로 구조를 짭니다🤔
***
### 조언 요청에 대한 저의 의견 
1. status code의 정상범위를 명시하는 방법
- 저는 결론적으로는 상수로 지금 구현하신것 처럼 빼는거에 한표 하겠습니다🙋🏻
최대한 실수를 줄일 수 있는 좋은 방법이라고 생각합니다!
200..<300은 통신을 한다면 중복되는 부분이 많을것이고 뭐 그럴일은 없겠다만..
만약 저 scope를 변경시켜야한다면 전역적으로 상수로 빼두셨다면 한 부분만 보수하면 될것이기에 유지보수 측면에서도 더 좋겠네요👍
2. unowned vs weak사용법
- 결론적으로 말씀드리자면 현재 코드 구현에서는 둘다 괜찮습니다.
먼저 확실히 짚고 넘어가봐야할 부분은 지금 클로저보다 다른 인스턴스의 수명이 확실히 긴지 동일한지 짧은지부터 봐야할것 같습니다.
(지금 짐작하고 계신게 잘못되어서 다시 생각해봐야한다는 의미는 아닙니다🥲)
아시겠지만 기존에는 weak와 unwoned의 대표적인 차이는 옵셔널 타입으로 선언해야하는지 입니다. (Swift 5이전)
이에 weak는 도중에 참조하고 있는 인스턴스가 메모리에서 빠져도 스스로 메모리 해제됩니다.
반면 unowned는 참조하고 있는 인스턴스가 존재하고 있는 도중에 해제되면 안됩니다.
그렇기에 말씀하신것처럼 self가 클로저보다 생명이 길때 두 방법 모두 이상 없습니다.
다만 두분의 질문으로 돌아온다면 확실하게 참조하는 인스턴스보다 오래 산다는걸 확신하고 보장할 수 있다면 unowned을 써도 됩니다.
메모리에서 두 옵션을 바로 봤을때의 차이는
weak는 메모리 해제 시 스택 주소값에 nil을 할당함으로 약한참조가 됩니다.
unowned는 메모리 해제 시 스택에 nil 할당이 아닌 dangler point(걍 동떨어져 있는 텅 빈 공간을 가르키는 포인터 입니다)가 남게됩니다.
그렇기에 앱 크래쉬를 발생시킬 수 있죠.
위에서 swift5 이전에는 옵셔널 타입을 가지고 두 차이를 구분했었는데 swift5 부터는 unowned도 옵셔널 타입을 지원해줍니다.
이와 별개로 여기서 핵심적으로 봐야할 부분은 여러분들이 알고 있는 인스턴스 해제 시점입니다.
그렇기에 서론에서 짚고 넘어가야한다는 부분을 다시 한번 생각해보면 좋을것 같아요.
여러분들이 가정하시기에 클로저 보다 타 인스턴스가 더 생명이 긴데 그럼 왜 지금의 weak 사용으로 문제가 없을까요?
이 부분도 한번 생각해보면 좋을것 같아요!
unowned로 변경하신다면 옵셔널 타입의 차이가 있기에 구현부에서 `self?`를 `self`로만 바꿔주면 되겠네요!
3. API URL Namespace 에 대한 조언
- 지금 구현 아주 제 스타일입니다👍
최대한 String Literal되는것들을 직접 사용하기보다 이렇게 꼭 파일과 타입으로 분리해서 관리하고 또 지금 결국 url 연산 프로퍼티를 구성해 그것만으로 접근하고 할 수 있도록 한 점도 너무 좋고 저도 이렇게 사용합니다!
아주 마이너한 의견으로는 리뷰 코멘트에도 남겨봤지만 해당 타입의 파일과 구조 위치에 대해 질문을 남겨봤습니다🙋🏻
***
여러분 이번주 오픈마켓 처음 접하시고 쉽지 않은 프로젝트 였을텐데 너무 고생 많으셨습니다!
고생하셨던 흔적이 너무 여실히 보여서 더 완성도가 높아졌던것 같아요👍
같이 의견을 나눠볼 포인트에 대해 리뷰를 남겨봤어요!
아! 요구사항들은 충족해서 다음 스텝으로 넘어가는데 무리는 없습니다.
만약 수정하신다면 수정 후 리뷰 재요청을 아니라면 머지 요청을 해주세요~!🥳

junbangg committed:
refactor: Extensions 폴더 가장 상위(모델과 동일한) 레벨로 이동 #1

na-young-kwon committed:
refactor: JOSNError타입 삭제 #1

junbangg committed:
refactor: APIError 리팩토링 #1

- 에러명 수정
- description 추가
- performDataTask 메서드에서 에러 헨들링 로직 변경

na-young-kwon committed:
refactor: 메서드의 인덴트 및 에러타입 이름 수정 #1

- Parsable -> JSONParsable 으로 에러타입 이름 변경
- MarketAPIService 테스트에서 에러를 비교하기 위해 Equatable 프로토콜 채택
- 파라미터가 긴 메서드의 인덴트 수정

junbangg committed:
refactor: APIServicable의 메서드명 변경 #1

na-young-kwon committed:
refactor: Page와 Product 타입의 프로퍼티들 한줄로 표현 #1

junbangg committed:
refactor: 테스트 더블 타입들 리팩토링 #2

- 각 타입명을 Mock에서 Stub 으로 변경
- 변경되지않는 프로퍼티 let 으로 변경
- 의미없는 줄바꿈 제거
- StubRequest의 JSON 데이터 형식 맞춤

na-young-kwon committed:
test: 파싱테스트 리팩토링 #2

- ParsableTest -> JSONParsableTests로 테스트명 변경
- 안전한 테스트를 위해 sut를 optional 타입으로 변경

junbangg commented:
너무 정성스러운 리뷰 감사합니다!
그린이 써주신 부분들에 대한 저희의 생각을 따로 적어봤어요!!

# 그린의 답변에 대한 답변!

- 파생 타입을 따로 가져가지 않은 이유
    - 메서드안에서 JSONDecoder를 만들기  vs  MarketAPIService가 인스턴스를 가지고있기 중에 고민함
    - 둘 중 전자를 선택한 이유는 MarketAPIService가 `JSON 파싱을 담당하는 타입`을 가지고 있으면 의존성이 높아진다고 생각 했습니다.
    - 타입을 따로 만들었을때의 장점이 생각나지 않습니다. 오히려 의존성이 높아진다 같은 단점만 생각이 나는데 혹시 Green이 생각하시는 이점이 있을까요??
    
- 알려주신 링크 참고하면서 `stub`, `mock` 의 차이에 대해서 공부를 같이 해봤습니다!
    - 저희가 이해한대로라면, 저희가 만든 `MockURLSession` , `MockURLSessionDataTask` 는 저희가 하드코딩한 “상태” 를 리턴하는 역할을 해주고있기 때문에 `Stub` 의 개념이 맞는것 같습니다.
    - 그리고 `sut` 로 정해줬던 `MarketAPIService` 에 대한 `Test Double` 을 만들었었다면 그건 `Mock` 이 맞는것 같다는 생각을 해봤습니다!
    - 네이밍을 적절하게 `stub` 으로 변경해보겠습니다!
    
    stub = 호출이 되면 미리 준비된 답변으로 응답하는것
    
    (일반적으로 우리가 mock으로 잘못 알고있는 test double) / 상태검증
    
    mock = 다른 테스트 더블과 달리 `행위` 를 검증하기 위한 테스트 더블
    
    행위검증을 추구한다는 것 자체가 다른 테스트 더블과의 큰 차이점 인 것 같습니다
    
    행위검증이란 `특정 메서드의 호출` 을 검증하는 것을 뜻합니다. 상태검증을 하기 어려운 경우 행위검증을 한다
    

- Protocol의 위치
    - 저희도 protocol의 위치를 고민했었는데 그린이 짐작하신대로 모델에서만 사용되기때문에 Model안에 위치시켰습니다.
    
- MarketAPIService의 해제시점? 클로저 보다 타 인스턴스가 더 생명이 긴데 그럼 왜 지금의 weak 사용으로 문제가 없을까요?
    - self(MarketAPIService)는 viewController가 살아있는 한 없어지지 않기 때문에 클로저보다 수명이 길다고 생각했습니다.
    - 따라서 인스턴스가 클로저보다 더 생명이 길기때문에 `weak`사용이 문제가 없다고 생각했습니다(`unowned` 도 문제가 없고). 그린이 말씀해주신 포인트를 잘 이해하지 못한것 같은데, 혹시 저희가 놓친 포인트가 있을까요?

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -95,17 +106,19 @@ extension MarketAPIService: APIServicable {
+                  }
             guard successRange.contains(statusCode) else {
-                completionHandler(.failure(APIError.unsuccessfulStatusCode(statusCode: statusCode)))
+                completionHandler(.failure(APIError.unsuccessfulStatusCode(statusCode: 410)))
```
410 에러 코드를 명시적으로 이 코드에 박아준 이유가 있나요~?🙋🏻

na-young-kwon replied:
저희가 `performDataTask` 메서드를 수정하면서 잠시 확인용으로 써놨던 에러인데 
이상태로 커밋이 되었는지는 몰랐네요ㅎㅎ
`APIError.unsuccessfulStatusCode(statusCode: statusCode)`로 수정해주었습니다!

GREENOVER reviewed (approved):
두분 다소 많은 코멘트를 드렸는데 답변도 다 잘해주시고 수정도 잘해주셔서 너무 감사합니다👍
코멘트에 대한 재 답변도 달아봤습니다ㅎㅎ
다음 스텝으로 넘어가기에 충분하다고 판단되어 마지막에  PR 메시지로 질문 주신것들만 요기 한번 정리해보고 머지하겠습니다!
***
1. 파생 타입을 따로 가져가지 않은 이유
- 이 부분은 제가 이해가 잘 안되서 다시 여쭤봤던 부분인것 같네요..!
상세한 설명 감사합니다ㅎㅎ 어떤 의도이셨는지 이해했습니다!
제가 생각하는 이점이라.. 흠 저는 그렇다면 타입으로 따로 분리를 한다면 말씀하신 의존성과 결합도 이 측면을 볼것 같아요.
사실 응집도는 뭐 동일한 역할을 동일하게 수행하는것임으로 패스하고,
정말 다 상황과 코드 구성에 따라 다르겠지만 의존성으로 주입받는것이 더 이점이 될 수도 있을것 같아요.
테스터블하기엔 그게 좀더 낫지 않을까란 생각도 해봤고 또 분리한다는것 자체가 결합도를 줄일 수 있을것 같다고 생각했어요.
2. weak vs unowned
- 아하 제가 궁금하고 질문을 드렸던 부분은 생명주기 관해서였습니다.
말씀하신대로  클로저보다 생명주기가 길것 같아 unowned을 택하는 질문을 주셨는데 그렇다면
weak의 사용은 타 인스턴스의 생명주기가 더 짧을때 사용이 될텐데 현재 코드에서 이상이 없는 이유를 여쭤봤습니다🙋🏻
***
혹시 제 답변들이나 의문사항이 이해가 되지 않는다면 바로 말씀해주세요!
다시 한번 고생많으셨고 만약 수정이 필요한 사항들이 있다면 다음 스텝에서 같이 녹여보면 될것 같습니다!
스텝1 고생많으셨습니다👍
스텝2도 화이팅입니다🙌
